pyccel/codegen/printing/ccode.py

```diff
@@ -1070,11 +1070,13 @@ def _print_PyccelDiv(self, expr):
 
     def _print_PyccelFloorDiv(self, expr):
         self._additional_imports.add("math")
-        if all(a.dtype is NativeInteger() for a in expr.args):
-            args = [PythonFloat(a) for a in expr.args]
-        else:
-            args = expr.args
-        code = ' / '.join(self._print(a) for a in args)
+        # the result type of the floor division is dependent on the arguments
+        # type, if all arguments are integers the result is integer otherwise
+        # the result type is float
+        need_to_cast = all(a.dtype is NativeInteger() for a in expr.args)
+        code = ' / '.join(self._print(a if a.dtype is NativeReal() else PythonFloat(a)) for a in expr.args)
+        if (need_to_cast):
+            return "(int64_t)floor({})".format(code)
```

Comment:
> the result type of the floor division is dependent on the arguments type, if all arguments are integers the result is integer otherwise the result type is float

You then convert all the arguments to reals (to handle negative values (`-3//4`->`-1` but in c `-3/4`->`0`) ) so doesn't the result always need to be cast?

Reply:
yes, you are right, we need to cast only one argument.

Reply:
Never mind I've just realised I was reading this the wrong way round. I forgot that the return type of `//` depends on the argument types.

Presumably `need_to_cast` is equivalent to `expr.dtype is NativeInteger`?